pkg/operator/starter.go

```diff
@@ -187,3 +202,35 @@ func RunOperator(ctx context.Context, controllerConfig *controllercmd.Controller
 
 	return nil
 }
+
+// getOperatorSyncState returns the management state of the operator to determine
+// how to sync conditional resources. It returns one of the following states:
+//
+//	Managed: resources should be synced
+//	Unmanaged: resources should NOT be synced
+//	Removed: resources should be deleted
+//
+// Errors fetching the operator state will log an error and return Unmanaged
+// to avoid syncing resources when the actual state is unknown.
+func getOperatorSyncState(operatorClient v1helpers.OperatorClientWithFinalizers) opv1.ManagementState {
+	opSpec, _, _, err := operatorClient.GetOperatorState()
+	if err != nil {
+		klog.Errorf("Failed to get operator state: %v", err)
+		return opv1.Unmanaged
+	}
+	// return the state from the operator if it's not managed
+	if opSpec.ManagementState != opv1.Managed {
+		return opSpec.ManagementState
```

Comment:
This will allow users to set `managementState: Removed` for CSI drivers that are not removable and the operator **will** remove.

That's an interesting feature that would allow users to disable e.g. vSphere CSI driver on clusters that don't want it installed. But I think it should be discussed first.

For now and for non-removable CSI drivers, `Removed` should be handled as `Managed` (?) and should throw some sane error. Returning `err` would degrade the cluster. It may be too strong, but it's the easiest thing to do.

Reply:
Ack, I reworked this in a second commit to maintain the existing behavior.

[NewStaticResourcesController](https://github.com/openshift/library-go/blob/eb2f24c36528fe928b6611b2bf88a734e346fcb6/pkg/operator/staticresourcecontroller/static_resource_controller.go#L117) sets the conditional resource functions to `nil` by default, which is interpreted by [WithConditionalResources](https://github.com/openshift/library-go/blob/eb2f24c36528fe928b6611b2bf88a734e346fcb6/pkg/operator/staticresourcecontroller/static_resource_controller.go#L146-L153) as "always create, never delete". So it's perfectly valid to pass `nil` functions into [WithConditionalStaticResourcesController](https://github.com/openshift/library-go/blob/eb2f24c36528fe928b6611b2bf88a734e346fcb6/pkg/operator/csi/csicontrollerset/csi_controller_set.go#L109-L110) and it will behave exactly as `WithStaticResourcesController` did.

So now removable operators set the conditional functions, and non-removable operators leave them `nil` for the simple static resources controller they always used before.

For removable operators, I changed `getOperatorSyncState` to be consistent with how we manage [static resources in gcp-filestore](https://github.com/openshift/gcp-filestore-csi-driver-operator/blob/2da7613226bf6d36240494484b816c72604c3ad1/pkg/operator/staticresources/staticresources.go#L114). It will not sync the resources if there is an error or if the state is not `Managed`, and it removes them only if `DeletionTimestamp` is set.